### PR TITLE
rework self-hosted compiler for incremental builds

### PR DESCRIPTION
**Pull request by andrewrk**
* introduce std.ArrayListUnmanaged for when you have the allocator
   stored elsewhere
 * move std.heap.ArenaAllocator implementation to its own file. extract
   the main state into std.heap.ArenaAllocator.State, which can be
   stored as an alternative to storing the entire ArenaAllocator, saving
   24 bytes per ArenaAllocator on 64 bit targets.
 * std.LinkedList.Node pointer field now defaults to being null
   initialized.
 * Rework self-hosted compiler Package API
 * Delete almost all the bitrotted self-hosted compiler code. The only bit
   rotted code left is in main.zig and compilation.zig
 * Add call instruction to ZIR
 * self-hosted compiler ir API and link API are reworked to support
   a long-running compiler that incrementally updates declarations
 * Introduce the concept of scopes to ZIR semantic analysis
 * ZIR text format supports referencing named decls that are declared
   later in the file
 * Figure out how memory management works for the long-running compiler
   and incremental compilation. The main roots are top level
   declarations. There is a table of decls. The key is a cryptographic
   hash of the fully qualified decl name. Each decl has an arena
   allocator where all of the memory related to that decl is stored.
   Each code block has its own arena allocator for the lifetime of
   the block. Values that want to survive when going out of scope in
   a block must get copied into the outer block. Finally, values must
   get copied into the Decl arena to be long-lived.
 * Delete the unused MemoryCell struct. Instead, comptime pointers are
   based on references to Decl structs.
 * Figure out how caching works. Each Decl will store a set of other
   Decls which must be recompiled when it changes.

This branch is still work-in-progress; this commit breaks the build.

### CODE REVIEW

File: src-self-hosted/ir.zig [hunk 1]
```diff
@@ -153,236 +167,711 @@ pub const Inst = struct {
...
-    exports: std.ArrayList(Module.Export),
-    fns: std.ArrayList(Module.Fn),
-    target: Target,
-    link_mode: std.builtin.LinkMode,
-    optimize_mode: std.builtin.Mode,
-    output_mode: std.builtin.OutputMode,
+    pub fn totalErrorCount(self: *Module) usize {
+        return self.failed_decls.size +
+            self.failed_fns.size +
+            self.failed_decls.size +
```
Comment (Vexu) [inline src-self-hosted/ir.zig:519]: This should probably be `failed_files`.

Comment (andrewrk) [reply src-self-hosted/ir.zig:519]: nice catch :+1:

**Comment by andrewrk** [comment]
60 second demo of this branch: https://andrewkelley.me/temp/zir-incremental.mp4

**Comment by FireFox317** [comment]
This library might be of good use for the codegen part of the self-hosted compiler: https://github.com/momumi/x86-zig

**Comment by andrewrk** [comment]
The leftover checkboxes are cards on the [self-hosted project board](https://github.com/ziglang/zig/projects/2). Once the self-hosted project gets further along, I'll convert all those cards into contributor friendly issues.

**Pull request merged**